commands/pipelines/promote.js

```diff
@@ -10,40 +13,61 @@ module.exports = {
   help: "promote an app's slug down the pipeline",
   needsApp: true,
   needsAuth: true,
-  flags: [
-    {name: 'stage', char: 's', description: 'stage of first app in pipeline', hasValue: true}
-  ],
   run: cli.command(function* (context, heroku) {
-    var stage;
+    const app = context.app;
 
-    stage = context.flags.stage;
-    let promise = heroku.request({
+    const coupling = yield cli.action(`Fetching app info`, heroku.request({
       method: 'GET',
-      path: `/apps/${context.app}/pipeline-couplings`,
-      headers: { 'Accept': 'application/vnd.heroku+json; version=3.pipelines' }
-    });
-    let coupling = yield cli.action(`Fetching app info`, promise);
+      path: `/apps/${app}/pipeline-couplings`,
+      headers: { 'Accept': PIPELINES_HEADER }
+    }));
 
-    promise = heroku.request({
-      method: 'GET',
-      path: `/pipelines/${coupling.pipeline.id}/apps`,
-      headers: { 'Accept': 'application/vnd.heroku+json; version=3.pipelines' }
-    }); // heroku.pipelines(pipeline_id).apps;
-    let apps = yield cli.action(`Fetching apps from ${coupling.pipeline.id}`, promise);
-
-    //let order = ["review", "development", "test", "qa", "staging", "production"];
-
-    let next_apps = [];
-    for (var app in apps) {
-      if (apps.hasOwnProperty(app)) {
-        if (apps[app].coupling.stage == stage) next_apps.push(app);
-      }
+    const allApps = yield cli.action(`Fetching apps from ${coupling.pipeline.name}`,
+      heroku.request({
+        method: 'GET',
+        path: `/pipelines/${coupling.pipeline.id}/apps`,
+        headers: { 'Accept': PIPELINES_HEADER }
+      }));
+
+    const sourceStage = coupling.stage;
+    const targetStage = PROMOTION_ORDER[PROMOTION_ORDER.indexOf(sourceStage) + 1];
+
+    if (targetStage == null) {
+      throw new Error(`Cannot promote ${app} from '${sourceStage}' stage`);
     }
-    cli.debug(next_apps);
-    for (var app in next_apps) {
-      if (keys.hasOwnProperty(app)) {
-        cli.log(`Promoting ${context.app} to ${app.name} (${stage})... done, v2`);
-      }
+
+    const targetApps = allApps.filter(function(app) {
+      return app.coupling.stage === targetStage;
+    });
+
+    const releases = yield cli.action(`Fetching latest release from ${app}`,
+      heroku.apps(app).releases().list({
+        headers: {
+          'Accept': V3_HEADER,
+          'Range':  'version ..; order=desc'
+        }
+      }));
```

Comment:
Does this do anything? I think .list() just takes a callback. When I cli.debug the releases they certainly don't look sorted and you follow up with a manual sort anyway...